### ISSUE TITLE
Improving perceived DOI registration/publicizing speed.

### ISSUE THREAD

**Issue by qqmyers** [IQSS/dataverse#7393]
With file PIDs enabled, and with large datasets (100s to 1000s of files), the time to register files at creation/upload (implemented in #7334 ), and to publicize those DOIs during publication, is large enough to be noticeable (simple tests on test servers show something like 2 DOI changes per second so ~10 minutes for 1K files). Given realistic file sizes, it isn't clear whether this is that significant during uploads (presumably 1000 'average' files take longer than 10 minutes to upload for most people?), but it is probably a big part of the wait at publication. 

I'm opening this issue to consider options for how the real or perceived performance might be improved. One option would be to make the changes asynchronously, e.g. the way that archiving can be done in a post-publication workflow. This would result in the user seeing file DOIs appear over time, or becoming public over time rather than being assigned/public when the save/publish operation completes. (Actually, since it takes significant time for DataCite to push newly publicized DOIs out to its index servers, there's already a delay between publish and the DOIs being resolvable).

It may also be possible to speed submission by parallelizing the calls to DataCite, or, if those need to be throttled and the delay is partially in Dataverse preparing the metadata to send, just parallelizing the prep steps and putting the DataCite calls in a pool.

Some of things that would help prioritize and refine this issue would be to know how much of a concern this is to installations and to know whether others have experience in sending large numbers of DOI requests to DataCite and or know what performance DataCite supports and whether they have bulk/asynchronous calls we could use.

I may investigate some, but this issue may sit for a while without community input.

**Comment by djbrooke** [comment]
Thanks @qqmyers for creating the issue. 

@scolapasta and @landreev I'm moving this to Needs Discussion so that we can... discuss in a tech hour. I'd like to figure this out before merging these two PRs:

- https://github.com/IQSS/dataverse/pull/7384
- https://github.com/IQSS/dataverse/pull/7334

We have messaging built up at publish time to handle a potential long delay around reserving file PIDs, but if we move the File PID reservation up in the workflow we could see a situation where there's not delay-related messaging and a long wait.

**Comment by qqmyers** [comment]
FWIW: Had an exchange with DataCite on this: 
As a developer working on Dataverse (https://github.com/IQSS/dataverse), I’d like to understand how we should interact with the DataCite API, particularly when performance is an issue. Specifically, it is an option in Dataverse to assign a DOI to individual files within a Dataset, and a Dataset may contain more than 1K files. Dataverse currently tries to register/publish all the DOIs for a Dataset and its DataFiles when a user presses the ‘publish’ button. (We’re moving to register file DOIs when files are uploaded and only adjust metadata/make the DOI findable at publication time.)
 
The general performance we're seeing in Dataverse is that we can register or publicize ~2 DOIs per second. That's reasonable performance for most uses, but it adds ~10 minutes to the time required to upload or publish Datasets with 1000 files (and we have real cases with at least double that number in the community). 
 
We’re considering a model where DOI updates would be made asynchronously in our code, i.e. the user would click ‘publish’ (or ‘upload’ in the future) and we’d show a success message with a note that DOIs will become public (or just registered for uploads) in the near future, at which point we can start running through a queue of needed changes.
 
We’d like to understand whether we need to do this only for larger datasets / as a backup if DataCite is busy/down or whether we should do this all the time. That depends somewhat on the performance we can get when making DOI changes. 
 
Some of the required time now (~1/2 sec per DOI) is certainly in Dataverse itself, so it would be useful to know what performance we could/should expect from DataCite in terms of throughput. Do you have such information? 
Is there a way to increase the speed of multiple operations? 
Can we parallelize our calls or would that cause server load issues at DataCite?
Are there any bulk or asynchronous APIs available/planned that would help? (Most of our uses cases would be for a batch of DOI changes to happen during dataset creation or publication, but the load itself would be fairly intermittent from any give Dataverse instance.)
If we move to asynchronous changes, is there a base rate we should target? 
Is there a way we could/should detect if DataCite is under load and delay/throttle updates?

Their response:
So the current status is that we do not impose per account limits, however there is a top level hard limit imposed by our firewall which is based on IP, this is around 3000 requests in a 5 minute window. We are looking to potentially implement account rate limits in the future.

To try and go through your specific questions
1.	We don't have specific performance metrics for registering DOIs, so we cant give you a guarantee on throughput.
2.	One way to improve performance is to make less API calls, if you use the REST API you can make 1 call to register both the metadata and the DOI as opposed to the MDS API which needs 2.
3.	You can send calls in parallel being mindful of the top level rate limit, its preferably not to hammer our servers however.
4.	We have no specific plans for bulk APIs at present however it has been discussed occasionally, so it's not beyond possibility.
5.	If you move to async, then as I said before, we don't have specific numbers, but bearing in mind our rate limit, I'd perhaps try starting with 5 requests per sec, if they are just all in one call DOI registrations, this would reduce your time down to a few minutes, however because we dont have throughput information this is not guaranteed.
6.	If you go over the rate limit, you will not be able to send any more DOIs. At this time we don't have anything more user friendly, though the idea of rate limiting per account would then then allow us to give better responses.

So -it seems like we should be able to increate our rate if we can parallelize, we should check the use of mds vs rest api to see if we're doing as few calls as possible, and we have an overall rate limit to handle.

**Comment by zbenta** [comment]
Hello there, any news on this topic?
We are trying to do a stress test to our infrastructure and we are using jmeter to do api calls to dataverse.
We are facing a similar situation, were we have 100 users creating datasets and uploading 2 small files to each dataset and we are having wait times of about 600 seconds since the moment the user created the dataset and uploaded the files.
We suspect that datacite is the culprit and this issue made us think we are right on our suspicions.

**Comment by pdurbin** [comment]
@zbenta not really. I'm somewhat surprised you're seeing performance problems with that scenario. 600 seconds? 10 minutes? That's super slow. I guess with jmeter you could test with the FAKE DOI provider to see if the slowness is from DataCite or not, which I sort of doubt (and @qqmyers agrees, from a quick conversation in Slack).

We do some perf testing with locust before release: https://guides.dataverse.org/en/5.13/developers/testing.html#locust

But I'm not sure if they cover your scenario. Anyway, if it turns out to not be a DataCite thing, please feel free to open a new issue.

**Comment by zbenta** [comment]
@pdurbin we tried going back to the FAKE DOI, removed al the settings of our account and still, we could see comunication between our dataverse instance and datacite.

Here you can see the output:

![image](https://user-images.githubusercontent.com/54443719/228486101-07e5106a-ac1c-4a71-b773-8e3aa778fd64.png)

We tried opening the resulting pcap file in wireshark, but the comunication is encrypted and we can't see what are try "chatting" about.

![image](https://user-images.githubusercontent.com/54443719/228486660-b3aed482-5bf7-49dd-9e3b-17b20338415e.png)

**Comment by pdurbin** [comment]
@zbenta I believe we got FAKE working for you a few days ago: https://matrix.to/#/!AmypvmJtUjBesRrnLM:matrix.org/$lt7_z6ks7msSTbPmJ-IsM5h2erKSLBbrkDtG-awFD8w?via=matrix.org&via=bertuch.name&via=sztaki.hu

**Comment by zbenta** [comment]
Yes @pdurbin , thanks for the tip.

**Comment by qqmyers** [comment]
Low priority now that file PIDs are optional/probably won't be used for datasets with large numbers of files.
